common/src/main/java/org/valkyrienskies/mod/mixin/client/world/MixinClientLevel.java

```diff
@@ -16,23 +17,41 @@
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
-import org.valkyrienskies.core.game.ships.QueryableShipDataImpl;
 import org.valkyrienskies.core.game.ships.ShipObjectClient;
-import org.valkyrienskies.core.game.ships.ShipObjectClientWorld;
-import org.valkyrienskies.mod.common.IShipObjectWorldClientProvider;
+import org.valkyrienskies.core.networking.VSNetworking;
 import org.valkyrienskies.mod.common.VSGameUtilsKt;
 import org.valkyrienskies.mod.common.util.EntityDragger;
 
 @Mixin(ClientLevel.class)
 public abstract class MixinClientLevel {
+    @Shadow
+    @Final
+    private Minecraft minecraft;
     @Shadow
     @Final
     private Int2ObjectMap<Entity> entitiesById;
 
+    @Unique
+    private boolean serverNoUdp = false;
+    @Unique
+    private int tryConnectIn = 100;
+
     @Inject(method = "tick", at = @At("HEAD"))
     private void preTick(final BooleanSupplier shouldKeepTicking, final CallbackInfo ci) {
         // Drag entities
         EntityDragger.Companion.dragEntitiesWithShips(entitiesById.values());
+        if (!VSNetworking.INSTANCE.getClientUsesUDP() && !serverNoUdp) {
```

Comment:
Can we move this logic to ShipObjectWorldClient?

Reply:
Hmm might be better yes

Reply:
Fixed but became also quite.. ugly